controllers/progressivesync_controller.go

```diff
@@ -60,171 +61,69 @@ type ProgressiveSyncReconciler struct {
-		}
-		r.Log.V(1).Info("apps selected", "apps", utils.GetAppsName(apps))
-
-		// Remove the annotation from the OutOfSync Applications before passing them to the Scheduler
-		// This action allows the Scheduler to keep track at which stage an Application has been synced.
-		outOfSyncApps := utils.GetAppsBySyncStatusCode(apps, argov1alpha1.SyncStatusCodeOutOfSync)
-		if err := r.removeAnnotationFromApps(ctx, outOfSyncApps, utils.ProgressiveSyncSyncedAtStageKey); err != nil {
-			message := "unable to remove out-of-sync annotation"
-			log.Error(err, message)
-			if err := r.updateStageStatus(ctx, stage.Name, message, syncv1alpha1.PhaseUnknown, &pr); err != nil {
-				return ctrl.Result{}, err
-			}
+	// Add finalizer if it doesn't exist
+	if !controllerutil.ContainsFinalizer(&pr, syncv1alpha1.ProgressiveSyncFinalizer) {
+		controllerutil.AddFinalizer(&pr, syncv1alpha1.ProgressiveSyncFinalizer)
+		if err := r.Update(ctx, &pr); err != nil {
+			log.Error(err, "failed to update object when adding finalizer")
 			return ctrl.Result{}, err
 		}
+		// Requeue after adding the finalizer
+		return ctrl.Result{Requeue: true}, nil
+	}
 
-		// Get the Applications to update
-		// TODO: update the scheduler to return the InProgress apps
-		scheduledApps := scheduler.Scheduler(apps, stage)
-
-		// TODO: status status - update targets
-
-		for _, s := range scheduledApps {
-			r.Log.Info("syncing app", "app", s)
-
-			_, err := r.syncApp(ctx, s.Name)
-
-			if err != nil {
-				if !strings.Contains(err.Error(), "another operation is already in progress") {
-					message := "unable to sync app"
-					log.Error(err, message, "message", err.Error())
-					if err := r.updateStageStatus(ctx, stage.Name, message, syncv1alpha1.PhaseUnknown, &pr); err != nil {
-						return ctrl.Result{}, err
-					}
-
-					// TODO: stage status - update failed clusters
+	for _, stage := range pr.Spec.Stages {
+		log = r.Log.WithValues("stage", stage.Name)
 
-					return ctrl.Result{}, err
-				}
-			}
+		pr, result, err := r.reconcileStage(ctx, pr, stage)
+		if err := r.updateStatusWithRetry(ctx, &pr); err != nil {
+			return ctrl.Result{}, err
 		}
 
-		if scheduler.IsStageFailed(apps) {
-			message := "stage failed"
-			r.Log.Info(message)
-			if err := r.updateStageStatus(ctx, stage.Name, message, syncv1alpha1.PhaseFailed, &pr); err != nil {
-				return ctrl.Result{}, err
-			}
-
-			failedMessage := fmt.Sprintf("%s stage failed", stage.Name)
-			failed := pr.NewStatusCondition(syncv1alpha1.CompletedCondition, metav1.ConditionTrue, syncv1alpha1.StagesFailedReason, failedMessage)
-			apimeta.SetStatusCondition(pr.GetStatusConditions(), failed)
-			if err := r.patchStatus(ctx, &pr); err != nil {
-				r.Log.Error(err, "failed to update object status")
-				return ctrl.Result{}, err
-			}
-			r.Log.Info("rollout failed")
-			// We can set Requeue: true once we have a timeout in place
-			return ctrl.Result{}, nil
+		if result.Requeue {
+			log.Info("requeuing stage")
```

Comment:
Nitpicking here, but wonder whether this is relevant as `Info` from a logging perspective?
This feels like `Debug` to me.

Reply:
Looking at the information we currently show at the `Info` level, it feels it gives you the full picture. We already say when an app is sync and what's the stage status.
I'm happy to move it to a different level if we thing it's more a debug info.

Reply:
we can still adjust it if the info logs become too verbose.